sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpError.cs

```diff
@@ -40,19 +30,194 @@ internal static class AmqpError
+        /// </summary>
+        ///
+        /// <param name="response">The response to consider.</param>
+        /// <param name="eventHubsResource">The Event Hubs resource associated with the request.</param>
+        ///
+        /// <returns>The exception that most accurately represents the response failure.</returns>
+        ///
+        public static Exception CreateExceptionForResponse(AmqpMessage response,
+                                                           string eventHubsResource)
+        {
+            if (response == null)
+            {
+                return new EventHubsException(true, eventHubsResource, Resources.UnknownCommunicationException);
+            }
+
+            if (!response.ApplicationProperties.Map.TryGetValue<string>(AmqpResponse.StatusDescription, out var description))
+            {
+                description = Resources.UnknownCommunicationException;
+            }
+
+            return CreateException(DetermineErrorCondition(response).Value, description, eventHubsResource);
+        }
+
+        /// <summary>
+        ///   Creates the exception that corresponds to a given AMQP error.
+        /// </summary>
+        ///
+        /// <param name="error">The AMQP error to consider.</param>
+        /// <param name="eventHubsResource">The Event Hubs resource associated with the operation.</param>
+        ///
+        /// <returns>The exception that most accurately represents the error that was encountered.</returns>
+        ///
+        public static Exception CreateExceptionForError(Error error,
+                                                        string eventHubsResource)
+        {
+            if (error == null)
+            {
+                return new EventHubsException(true, eventHubsResource, Resources.UnknownCommunicationException);
+            }
+
+            return CreateException(error.Condition.Value, error.Description, eventHubsResource);
+        }
+
+        /// <summary>
+        ///   Creates the exception that corresponds to a given AMQP failure scenario.
+        /// </summary>
+        ///
+        /// <param name="condition">The error condition that represents the failure scenario.</param>
+        /// <param name="description">The descriptive text to use for messaging the scenario.</param>
+        /// <param name="eventHubsResource">The Event Hubs resource associated with the failure.</param>
+        ///
+        /// <returns>The exception that most accurately represents the failure scenario.</returns>
+        ///
+        private static Exception CreateException(string condition,
+                                                 string description,
+                                                 string eventHubsResource)
+        {
+            // The request timed out.
+
+            if (String.Equals(condition, TimeoutError.Value, StringComparison.InvariantCultureIgnoreCase))
```

Comment:
Not in love with this implementation, but I wasn't able to come up with something more elegant.   I attempted to order the cases from most likely to least likely to be as efficient as possible, but want to also acknowledge that this is an error path only.